File: CarlaDashBoard/bridge_server/bridge/map_graph.py
"""Serializes a router_plan RouteGraph (planner.graph — node_id -> carla.Waypoint, node_id ->
list[Edge]) into the JSON shape design doc §04/§08 expects for GET /maps/{town}: a `nodes`
list and an `edges` list, same field names as map_nodes.csv/map_edges.csv so this reads the
same whether it came from a live graph or an exported CSV (see router_plan/README.md,
"Quyết định đã chốt", for why the graph itself is built live rather than round-tripped
through CSV).
"""


def build_payload(town, graph):
    nodes = []
    # `graph.nodes` anh xa node_id -> carla.Transform, KHONG phai carla.Waypoint:
    # router_plan/graph_builder.py co y chi giu Transform ("chi giu Transform, khong giu ca
    # Waypoint") de khong om hang nghin doi tuong Waypoint song. Vi vay road_id/section_id/
    # lane_id/s khong con doc duoc o day va bi bo khoi payload — MapCanvas ben WPF chi ve
    # tu x/y nen khong dung toi chung. Ban truoc doc `waypoint.transform.location` va nem
    # AttributeError, lam /maps/{town} tra 503 vinh vien (minimap trong).
    for node_id, transform in graph.nodes.items():
        loc = transform.location
        nodes.append({
            "node_id": node_id,
            "x": round(loc.x, 2),
            "y": round(loc.y, 2),
            "z": round(loc.z, 2),
            "yaw_deg": round(transform.rotation.yaw, 2),
        })

    edges = []
    for from_id, edge_list in graph.edges.items():
        for edge in edge_list:
            edges.append({
                "from_node_id": from_id,
                "to_node_id": edge.to_id,
                "edge_type": edge.edge_type,
                "cost_m": round(edge.cost_m, 2),
                "yaw_delta_deg": round(edge.yaw_delta_deg, 2),
            })

    return {
        "type": "MapGraph",
        "town": town,
        "resolution_m": graph.resolution_m,
        "node_count": len(nodes),
        "edge_count": len(edges),
        "nodes": nodes,
        "edges": edges,
    }

File: CarlaDashBoard/CarlaDashBoard.Wpf/Models/MapGraph.cs
using System.Text.Json.Serialization;

namespace CarlaDashBoard.Wpf.Models;

/// <summary>One node of the A* graph — same fields as map_graph.py's build_payload().</summary>
public sealed class MapNode
{
    [JsonPropertyName("node_id")] public string NodeId { get; set; } = "";
    [JsonPropertyName("road_id")] public int RoadId { get; set; }
    [JsonPropertyName("section_id")] public int SectionId { get; set; }
    [JsonPropertyName("lane_id")] public int LaneId { get; set; }
    [JsonPropertyName("s")] public double S { get; set; }
    [JsonPropertyName("x")] public double X { get; set; }
    [JsonPropertyName("y")] public double Y { get; set; }
    [JsonPropertyName("z")] public double Z { get; set; }
    [JsonPropertyName("yaw_deg")] public double YawDeg { get; set; }
}

public sealed class MapEdge
{
    [JsonPropertyName("from_node_id")] public string FromNodeId { get; set; } = "";
    [JsonPropertyName("to_node_id")] public string ToNodeId { get; set; } = "";
    [JsonPropertyName("edge_type")] public string EdgeType { get; set; } = "";
    [JsonPropertyName("cost_m")] public double CostM { get; set; }
    [JsonPropertyName("yaw_delta_deg")] public double YawDeltaDeg { get; set; }
}

/// <summary>Mirrors the JSON body of GET /maps/{town} (design doc §01/§04, bridge's map_graph.py).</summary>
public sealed class MapGraph
{
    [JsonPropertyName("town")] public string Town { get; set; } = "";
    [JsonPropertyName("resolution_m")] public double ResolutionM { get; set; }
    [JsonPropertyName("node_count")] public int NodeCount { get; set; }
    [JsonPropertyName("edge_count")] public int EdgeCount { get; set; }
    [JsonPropertyName("nodes")] public List<MapNode> Nodes { get; set; } = new();
    [JsonPropertyName("edges")] public List<MapEdge> Edges { get; set; } = new();
}
